File: harnessarena_uploader/trackers.py
"""Reusable tracker classes for harness session parsing.

Extracts the duplicated state machines from claude.py and codex.py into
composable, zero-dependency tracker objects. Each tracker owns one concern
(time spans, plan mode, daily breakdown, tool classification, subagents).
"""
from __future__ import annotations

from collections import Counter
from datetime import datetime
from typing import Optional

from .helpers import _register_mcp_tool
from .models import SubagentMeta


class TimeSpanTracker:
    """Track harness_exec and user_idle time spans across turns.

    State machine: on_user_turn() emits the previous turn's exec + idle spans,
    on_nonuser_event() updates the latest non-user timestamp, and on_tool_start/end
    track individual tool call durations within a turn.
    """

    def __init__(self) -> None:
        self._last_user_ts: Optional[datetime] = None
        self._last_nonuser_ts: Optional[datetime] = None
        self._pending_tool_calls: dict[str, dict] = {}
        self._current_turn_tool_spans: list[dict] = []
        self._current_turn_input_tokens: int = 0
        self._current_turn_output_tokens: int = 0
        self._time_spans: list[dict] = []
        self._turn_exec_times: list[float] = []

    def on_user_turn(self, dt: datetime) -> None:
        """Record a real user turn boundary.

        Emits harness_exec + user_idle spans for the previous turn if applicable.
        Resets per-turn state.
        """
        if self._last_user_ts and self._last_nonuser_ts:
            exec_dur = (self._last_nonuser_ts - self._last_user_ts).total_seconds()
            idle_dur = (dt - self._last_nonuser_ts).total_seconds()
            if exec_dur > 0:
                span: dict = {
                    "type": "harness_exec",
                    "start": self._last_user_ts.isoformat(),
                    "end": self._last_nonuser_ts.isoformat(),
                    "seconds": round(exec_dur, 1),
                }
                if self._current_turn_tool_spans:
                    span["tool_spans"] = self._current_turn_tool_spans[:]
                if self._current_turn_input_tokens > 0 or self._current_turn_output_tokens > 0:
                    span["input_tokens"] = self._current_turn_input_tokens
                    span["output_tokens"] = self._current_turn_output_tokens
                    span["tokens"] = self._current_turn_input_tokens + self._current_turn_output_tokens
                self._time_spans.append(span)
                if exec_dur < 1800:
                    self._turn_exec_times.append(exec_dur)
            if idle_dur > 0:
                self._time_spans.append({
                    "type": "user_idle",
                    "start": self._last_nonuser_ts.isoformat(),
                    "end": dt.isoformat(),
                    "seconds": round(idle_dur, 1),
                })
        self._last_user_ts = dt
        self._last_nonuser_ts = None
        self._current_turn_tool_spans = []
        self._current_turn_input_tokens = 0
        self._current_turn_output_tokens = 0
        self._pending_tool_calls = {}

    def on_nonuser_event(self, dt: datetime) -> None:
        """Record the timestamp of a non-user event (assistant, system, tool output)."""
        self._last_nonuser_ts = dt

    def on_tokens(self, input_tokens: int, output_tokens: int = 0) -> None:
        """Accumulate tokens for the current turn."""
        self._current_turn_input_tokens += input_tokens
        self._current_turn_output_tokens += output_tokens

    def on_tool_start(self, call_id: str, name: str, category: str, dt: datetime) -> None:
        """Record a tool call start for span timing."""
        self._pending_tool_calls[call_id] = {
            "name": name,
            "category": category,
            "start_dt": dt,
        }

    def on_tool_end(self, call_id: str, dt: datetime) -> Optional[dict]:
        """Record a tool call end. Returns the tool_span dict if matched, else None."""
        if call_id not in self._pending_tool_calls:
            return None
        pending = self._pending_tool_calls.pop(call_id)
        dur = (dt - pending["start_dt"]).total_seconds()
        if dur >= 0:
            tool_span_entry = {
                "name": pending["name"],
                "category": pending["category"],
                "start": pending["start_dt"].isoformat(),
                "end": dt.isoformat(),
                "seconds": round(dur, 3),
            }
            self._current_turn_tool_spans.append(tool_span_entry)
            return tool_span_entry
        return None

    def finalize(self) -> tuple[list[dict], list[float]]:
        """Capture the last turn and return (time_spans, turn_exec_times)."""
        if self._last_user_ts and self._last_nonuser_ts:
            exec_dur = (self._last_nonuser_ts - self._last_user_ts).total_seconds()
            if exec_dur > 0:
                span: dict = {
                    "type": "harness_exec",
                    "start": self._last_user_ts.isoformat(),
                    "end": self._last_nonuser_ts.isoformat(),
                    "seconds": round(exec_dur, 1),
                }
                if self._current_turn_tool_spans:
                    span["tool_spans"] = self._current_turn_tool_spans[:]
                if self._current_turn_input_tokens > 0 or self._current_turn_output_tokens > 0:
                    span["input_tokens"] = self._current_turn_input_tokens
                    span["output_tokens"] = self._current_turn_output_tokens
                    span["tokens"] = self._current_turn_input_tokens + self._current_turn_output_tokens
                self._time_spans.append(span)
                if exec_dur < 1800:
                    self._turn_exec_times.append(exec_dur)
        return self._time_spans, self._turn_exec_times


class PlanModeTracker:
    """Track plan mode enter/exit events and plan mode time spans."""

    def __init__(self) -> None:
        self._entries: int = 0
        self._exits: int = 0
        self._start: Optional[datetime] = None
        self._tool_spans: list[dict] = []
        self._spans: list[dict] = []

    def on_enter(self, dt: datetime) -> None:
        """Record plan mode entry."""
        self._entries += 1
        self._start = dt
        self._tool_spans = []

    def on_exit(self, dt: datetime) -> None:
        """Record plan mode exit. Emits a plan_mode span if timing is available."""
        self._exits += 1
        if self._start:
            dur = (dt - self._start).total_seconds()
            if dur > 0:
                ps: dict = {
                    "type": "plan_mode",
                    "start": self._start.isoformat(),
                    "end": dt.isoformat(),
                    "seconds": round(dur, 1),
                }
                if self._tool_spans:
                    ps["tool_spans"] = self._tool_spans[:]
                self._spans.append(ps)
            self._start = None
            self._tool_spans = []

    def on_tool_span(self, span: dict) -> None:
        """Record a tool span that occurred during plan mode."""
        if self._start is not None:
            self._tool_spans.append(span)

    def replace_session_level(self, time_spans: list[dict], plan_entries: int) -> None:
        """If session-level plan mode and no explicit plan spans, replace harness_exec with plan_mode.

        Modifies time_spans in place.
        """
        is_plan = (self._entries > 0 or plan_entries > 0)
        if not is_plan or not time_spans:
            return
        if any(s.get("type") == "plan_mode" for s in time_spans):
            return
        all_starts = [s["start"] for s in time_spans if s.get("type") == "harness_exec"]
        all_ends = [s["end"] for s in time_spans if s.get("type") == "harness_exec"]
        if not all_starts or not all_ends:
            return
        ps_start = min(all_starts)
        ps_end = max(all_ends)
        total_exec = sum(s["seconds"] for s in time_spans if s.get("type") == "harness_exec")
        plan_tool_spans = []
        for s in time_spans:
            if s.get("type") == "harness_exec" and "tool_spans" in s:
                plan_tool_spans.extend(s["tool_spans"])
        time_spans[:] = [s for s in time_spans if s.get("type") != "harness_exec"]
        plan_span_entry: dict = {
            "type": "plan_mode",
            "start": ps_start,
            "end": ps_end,
            "seconds": round(total_exec, 1),
        }
        if plan_tool_spans:
            plan_span_entry["tool_spans"] = plan_tool_spans
        time_spans.append(plan_span_entry)

    @property
    def entries(self) -> int:
        return self._entries

    @property
    def exits(self) -> int:
        return self._exits

    @property
    def spans(self) -> list[dict]:
        return self._spans

    @property
    def is_active(self) -> bool:
        return self._start is not None


class DailyTracker:
    """Accumulate per-date metric counters."""

    def __init__(self) -> None:
        self._data: dict[str, dict] = {}

    def add(self, date_str: str, **kwargs: int) -> None:
        if date_str not in self._data:
            self._data[date_str] = {
                "date": date_str, "tokens_in": 0, "tokens_out": 0, "tokens_total": 0,
                "prompts": 0, "sessions": 0, "subagent_calls": 0,
                "background_agents": 0, "tool_calls": 0, "mcp_calls": 0,
            }
        for k, v in kwargs.items():
            self._data[date_str][k] = self._data[date_str].get(k, 0) + v

    def mark_session_start(self, date_str: str) -> None:
        """Mark that a session started on this date (sets sessions=1)."""
        if date_str in self._data:
            self._data[date_str]["sessions"] = 1

    def finalize(self) -> list[dict]:
        return sorted(self._data.values(), key=lambda d: d["date"])


class ToolClassifier:
    """Classify tool calls and accumulate counters.

    Delegates to the parser's classify_tool_call method for detection,
    then updates all relevant counters and trackers as side effects.
    """

    def __init__(self) -> None:
        self._subagent_calls: int = 0
        self._background_agents: int = 0
        self._mcp_calls: int = 0
        self._tool_names: Counter = Counter()
        self._tool_categories: dict[str, str] = {}  # tool_name -> category
        self._skill_invocations: Counter = Counter()
        self._mcp_servers: dict[str, dict] = {}

    def record(
        self,
        tool_name: str,
        tool_input: dict,
        parser,
        timestamp_dt: Optional[datetime] = None,
        plan_tracker: Optional[PlanModeTracker] = None,
        time_tracker: Optional[TimeSpanTracker] = None,
        subagent_collector: Optional[SubagentCollector] = None,
        daily_tracker: Optional[DailyTracker] = None,
        date: Optional[str] = None,
    ) -> str:
        """Classify a tool call and update all counters/trackers.

        Returns the category: "tool", "subagent", "mcp", or "skill".
        """
        self._tool_names[tool_name] += 1
        category: str = "tool"

        if parser is None:
            # Fallback inline detection (no parser)
            if tool_name == "Agent":
                category = "subagent"
                self._subagent_calls += 1
                if daily_tracker and date:
                    daily_tracker.add(date, subagent_calls=1)
                if isinstance(tool_input, dict) and tool_input.get("run_in_background"):
                    self._background_agents += 1
                    if daily_tracker and date:
                        daily_tracker.add(date, background_agents=1)
            elif tool_name == "Skill":
                category = "skill"
                skill = tool_input.get("skill", "unknown") if isinstance(tool_input, dict) else "unknown"
                self._skill_invocations[skill] += 1
            elif tool_name.startswith("mcp__"):
                category = "mcp"
                self._mcp_calls += 1
                if daily_tracker and date:
                    daily_tracker.add(date, mcp_calls=1)
            return category

        c = parser.classify_tool_call(tool_name, tool_input)

        if c["is_subagent"]:
            category = "subagent"
            self._subagent_calls += 1
            if daily_tracker and date:
                daily_tracker.add(date, subagent_calls=1)
            is_bg = c["is_background_agent"]
            if is_bg:
                self._background_agents += 1
                if daily_tracker and date:
                    daily_tracker.add(date, background_agents=1)
            if subagent_collector:
                subagent_collector.record_spawn(
                    mode="background" if is_bg else "foreground",
                    subagent_type=tool_input.get("subagent_type", "") if isinstance(tool_input, dict) else "",
                    description=tool_input.get("description", "") if isinstance(tool_input, dict) else "",
                )

        if c["skill_name"]:
            category = "skill"
            self._skill_invocations[c["skill_name"]] += 1

        if c["is_mcp"]:
            category = "mcp"
            self._mcp_calls += 1
            _register_mcp_tool(self._mcp_servers, tool_name)
            if daily_tracker and date:
                daily_tracker.add(date, mcp_calls=1)

        if c["is_plan_enter"] and plan_tracker:
            if timestamp_dt:
                plan_tracker.on_enter(timestamp_dt)
            else:
                plan_tracker._entries += 1

        if c["is_plan_exit"] and plan_tracker:
            if timestamp_dt:
                plan_tracker.on_exit(timestamp_dt)
            else:
                plan_tracker._exits += 1

        self._tool_categories[tool_name] = category
        return category

    @property
    def tool_categories(self) -> dict[str, str]:
        return self._tool_categories

    @property
    def subagent_calls(self) -> int:
        return self._subagent_calls

    @property
    def background_agents(self) -> int:
        return self._background_agents

    @property
    def mcp_calls(self) -> int:
        return self._mcp_calls

    @property
    def tool_names(self) -> Counter:
        return self._tool_names

    @property
    def skill_invocations(self) -> Counter:
        return self._skill_invocations

    @property
    def mcp_servers(self) -> dict:
        return self._mcp_servers


class SubagentCollector:
    """Collect subagent metadata from tool_use blocks and directory listings."""

    def __init__(self) -> None:
        self._metas: list[dict] = []

    def record_spawn(self, mode: str = "foreground", subagent_type: str = "", description: str = "") -> None:
        self._metas.append({
            "ordinal": len(self._metas),
            "mode": mode,
            "subagent_type": subagent_type,
            "description": description,
        })

    def enrich(self, idx: int, total_tokens: int = 0, total_tool_calls: int = 0) -> None:
        """Enrich the subagent at index idx with per-file metrics."""
        if idx < len(self._metas):
            self._metas[idx]["total_tokens"] = total_tokens
            self._metas[idx]["total_tool_calls"] = total_tool_calls

    def ensure_count(self, min_count: int) -> None:
        """Fill placeholders if directory count > tool_use count."""
        while len(self._metas) < min_count:
            self._metas.append({
                "ordinal": len(self._metas),
                "mode": "foreground",
                "subagent_type": "",
                "description": "",
            })

    def finalize(self) -> tuple:
        """Return tuple of SubagentMeta from collected data."""
        return tuple(SubagentMeta(**m) for m in self._metas)

    @property
    def raw_metas(self) -> list[dict]:
        """Access the raw metas list (for Codex spawn_agent enrichment)."""
        return self._metas

    def __len__(self) -> int:
        return len(self._metas)

File: src/trackers.ts
import { registerMcpTool } from "./helpers.js";
import { createSubagentMeta, type SubagentMeta } from "./models.js";

// --- TimeSpanTracker ---

export interface TimeSpan {
  type: string;
  start: string;
  end: string;
  seconds: number;
  tool_spans?: ToolSpan[];
  input_tokens?: number;
  output_tokens?: number;
  tokens?: number;
}

export interface ToolSpan {
  name: string;
  category: string;
  start: string;
  end: string;
  seconds: number;
}

interface PendingToolCall {
  name: string;
  category: string;
  startDt: Date;
}

export class TimeSpanTracker {
  private lastUserTs: Date | null = null;
  private lastNonuserTs: Date | null = null;
  private pendingToolCalls = new Map<string, PendingToolCall>();
  private currentTurnToolSpans: ToolSpan[] = [];
  private currentTurnInputTokens = 0;
  private currentTurnOutputTokens = 0;
  private _timeSpans: TimeSpan[] = [];
  private _turnExecTimes: number[] = [];

  onUserTurn(dt: Date): void {
    if (this.lastUserTs && this.lastNonuserTs) {
      const execDur = (this.lastNonuserTs.getTime() - this.lastUserTs.getTime()) / 1000;
      const idleDur = (dt.getTime() - this.lastNonuserTs.getTime()) / 1000;
      if (execDur > 0) {
        const span: TimeSpan = {
          type: "harness_exec",
          start: this.lastUserTs.toISOString(),
          end: this.lastNonuserTs.toISOString(),
          seconds: Math.round(execDur * 10) / 10,
        };
        if (this.currentTurnToolSpans.length > 0) {
          span.tool_spans = [...this.currentTurnToolSpans];
        }
        if (this.currentTurnInputTokens > 0 || this.currentTurnOutputTokens > 0) {
          span.input_tokens = this.currentTurnInputTokens;
          span.output_tokens = this.currentTurnOutputTokens;
          span.tokens = this.currentTurnInputTokens + this.currentTurnOutputTokens;
        }
        this._timeSpans.push(span);
        if (execDur < 1800) {
          this._turnExecTimes.push(execDur);
        }
      }
      if (idleDur > 0) {
        this._timeSpans.push({
          type: "user_idle",
          start: this.lastNonuserTs.toISOString(),
          end: dt.toISOString(),
          seconds: Math.round(idleDur * 10) / 10,
        });
      }
    }
    this.lastUserTs = dt;
    this.lastNonuserTs = null;
    this.currentTurnToolSpans = [];
    this.currentTurnInputTokens = 0;
    this.currentTurnOutputTokens = 0;
    this.pendingToolCalls.clear();
  }

  onNonuserEvent(dt: Date): void {
    this.lastNonuserTs = dt;
  }

  onTokens(inputTokens: number, outputTokens = 0): void {
    this.currentTurnInputTokens += inputTokens;
    this.currentTurnOutputTokens += outputTokens;
  }

  onToolStart(callId: string, name: string, category: string, dt: Date): void {
    this.pendingToolCalls.set(callId, { name, category, startDt: dt });
  }

  onToolEnd(callId: string, dt: Date): ToolSpan | null {
    const pending = this.pendingToolCalls.get(callId);
    if (!pending) return null;
    this.pendingToolCalls.delete(callId);
    const dur = (dt.getTime() - pending.startDt.getTime()) / 1000;
    if (dur >= 0) {
      const toolSpan: ToolSpan = {
        name: pending.name,
        category: pending.category,
        start: pending.startDt.toISOString(),
        end: dt.toISOString(),
        seconds: Math.round(dur * 1000) / 1000,
      };
      this.currentTurnToolSpans.push(toolSpan);
      return toolSpan;
    }
    return null;
  }

  finalize(): { timeSpans: TimeSpan[]; turnExecTimes: number[] } {
    if (this.lastUserTs && this.lastNonuserTs) {
      const execDur = (this.lastNonuserTs.getTime() - this.lastUserTs.getTime()) / 1000;
      if (execDur > 0) {
        const span: TimeSpan = {
          type: "harness_exec",
          start: this.lastUserTs.toISOString(),
          end: this.lastNonuserTs.toISOString(),
          seconds: Math.round(execDur * 10) / 10,
        };
        if (this.currentTurnToolSpans.length > 0) {
          span.tool_spans = [...this.currentTurnToolSpans];
        }
        if (this.currentTurnInputTokens > 0 || this.currentTurnOutputTokens > 0) {
          span.input_tokens = this.currentTurnInputTokens;
          span.output_tokens = this.currentTurnOutputTokens;
          span.tokens = this.currentTurnInputTokens + this.currentTurnOutputTokens;
        }
        this._timeSpans.push(span);
        if (execDur < 1800) {
          this._turnExecTimes.push(execDur);
        }
      }
    }
    return { timeSpans: this._timeSpans, turnExecTimes: this._turnExecTimes };
  }

  // Expose for OpenCode parser direct access
  get lastNonuserTimestamp(): Date | null {
    return this.lastNonuserTs;
  }

  get pendingCalls(): Map<string, PendingToolCall> {
    return this.pendingToolCalls;
  }
}

// --- PlanModeTracker ---

export class PlanModeTracker {
  private _entries = 0;
  private _exits = 0;
  private _start: Date | null = null;
  private _toolSpans: ToolSpan[] = [];
  private _spans: TimeSpan[] = [];

  onEnter(dt: Date): void {
    this._entries++;
    this._start = dt;
    this._toolSpans = [];
  }

  onExit(dt: Date): void {
    this._exits++;
    if (this._start) {
      const dur = (dt.getTime() - this._start.getTime()) / 1000;
      if (dur > 0) {
        const ps: TimeSpan = {
          type: "plan_mode",
          start: this._start.toISOString(),
          end: dt.toISOString(),
          seconds: Math.round(dur * 10) / 10,
        };
        if (this._toolSpans.length > 0) {
          ps.tool_spans = [...this._toolSpans];
        }
        this._spans.push(ps);
      }
      this._start = null;
      this._toolSpans = [];
    }
  }

  onToolSpan(span: ToolSpan): void {
    if (this._start !== null) {
      this._toolSpans.push(span);
    }
  }

  replaceSessionLevel(timeSpans: TimeSpan[], planEntries: number): void {
    const isPlan = this._entries > 0 || planEntries > 0;
    if (!isPlan || timeSpans.length === 0) return;
    if (timeSpans.some((s) => s.type === "plan_mode")) return;

    const allStarts = timeSpans.filter((s) => s.type === "harness_exec").map((s) => s.start);
    const allEnds = timeSpans.filter((s) => s.type === "harness_exec").map((s) => s.end);
    if (allStarts.length === 0 || allEnds.length === 0) return;

    const psStart = allStarts.sort()[0];
    const psEnd = allEnds.sort().at(-1)!;
    const totalExec = timeSpans
      .filter((s) => s.type === "harness_exec")
      .reduce((sum, s) => sum + s.seconds, 0);
    const planToolSpans: ToolSpan[] = [];
    for (const s of timeSpans) {
      if (s.type === "harness_exec" && s.tool_spans) {
        planToolSpans.push(...s.tool_spans);
      }
    }

    // Remove harness_exec spans in-place
    const nonExec = timeSpans.filter((s) => s.type !== "harness_exec");
    timeSpans.length = 0;
    timeSpans.push(...nonExec);

    const planSpanEntry: TimeSpan = {
      type: "plan_mode",
      start: psStart,
      end: psEnd,
      seconds: Math.round(totalExec * 10) / 10,
    };
    if (planToolSpans.length > 0) {
      planSpanEntry.tool_spans = planToolSpans;
    }
    timeSpans.push(planSpanEntry);
  }

  get entries(): number { return this._entries; }
  set entries(v: number) { this._entries = v; }
  get exits(): number { return this._exits; }
  set exits(v: number) { this._exits = v; }
  get spans(): TimeSpan[] { return this._spans; }
  get isActive(): boolean { return this._start !== null; }
}

// --- DailyTracker ---

export interface DailyEntry {
  date: string;
  tokens_in: number;
  tokens_out: number;
  tokens_total: number;
  prompts: number;
  sessions: number;
  subagent_calls: number;
  background_agents: number;
  tool_calls: number;
  mcp_calls: number;
}

export class DailyTracker {
  private data = new Map<string, DailyEntry>();

  add(dateStr: string, updates: Partial<Omit<DailyEntry, "date">>): void {
    if (!this.data.has(dateStr)) {
      this.data.set(dateStr, {
        date: dateStr,
        tokens_in: 0, tokens_out: 0, tokens_total: 0,
        prompts: 0, sessions: 0, subagent_calls: 0,
        background_agents: 0, tool_calls: 0, mcp_calls: 0,
      });
    }
    const entry = this.data.get(dateStr)!;
    for (const [k, v] of Object.entries(updates)) {
      if (k !== "date" && typeof v === "number") {
        (entry as unknown as Record<string, unknown>)[k] =
          ((entry as unknown as Record<string, unknown>)[k] as number ?? 0) + v;
      }
    }
  }

  markSessionStart(dateStr: string): void {
    const entry = this.data.get(dateStr);
    if (entry) entry.sessions = 1;
  }

  finalize(): DailyEntry[] {
    return [...this.data.values()].sort((a, b) => a.date.localeCompare(b.date));
  }
}

// --- ToolClassifier ---

export interface ToolClassification {
  is_subagent: boolean;
  is_background_agent: boolean;
  is_mcp: boolean;
  skill_name: string | null;
  is_plan_enter: boolean;
  is_plan_exit: boolean;
}

export interface ClassifyToolCallFn {
  classifyToolCall(toolName: string, toolInput: Record<string, unknown>): ToolClassification;
}

export class ToolClassifier {
  private _subagentCalls = 0;
  private _backgroundAgents = 0;
  private _mcpCalls = 0;
  readonly toolNames = new Map<string, number>();
  readonly toolCategories = new Map<string, string>();
  readonly skillInvocations = new Map<string, number>();
  readonly mcpServers: Record<string, Record<string, unknown>> = {};

  record(
    toolName: string,
    toolInput: Record<string, unknown>,
    parser: ClassifyToolCallFn | null,
    opts?: {
      timestampDt?: Date;
      planTracker?: PlanModeTracker;
      timeTracker?: TimeSpanTracker;
      subagentCollector?: SubagentCollector;
      dailyTracker?: DailyTracker;
      date?: string;
    },
  ): string {
    this.toolNames.set(toolName, (this.toolNames.get(toolName) ?? 0) + 1);
    let category = "tool";

    if (!parser) {
      // Fallback inline detection
      if (toolName === "Agent") {
        category = "subagent";
        this._subagentCalls++;
        if (opts?.dailyTracker && opts.date) {
          opts.dailyTracker.add(opts.date, { subagent_calls: 1 });
        }
        if (toolInput.run_in_background) {
          this._backgroundAgents++;
          if (opts?.dailyTracker && opts.date) {
            opts.dailyTracker.add(opts.date, { background_agents: 1 });
          }
        }
      } else if (toolName === "Skill") {
        category = "skill";
        const skill = (typeof toolInput === "object" ? (toolInput.skill as string) : null) ?? "unknown";
        this.skillInvocations.set(skill, (this.skillInvocations.get(skill) ?? 0) + 1);
      } else if (toolName.startsWith("mcp__")) {
        category = "mcp";
        this._mcpCalls++;
        if (opts?.dailyTracker && opts.date) {
          opts.dailyTracker.add(opts.date, { mcp_calls: 1 });
        }
      }
      return category;
    }

    const c = parser.classifyToolCall(toolName, toolInput);

    if (c.is_subagent) {
      category = "subagent";
      this._subagentCalls++;
      if (opts?.dailyTracker && opts.date) {
        opts.dailyTracker.add(opts.date, { subagent_calls: 1 });
      }
      if (c.is_background_agent) {
        this._backgroundAgents++;
        if (opts?.dailyTracker && opts.date) {
          opts.dailyTracker.add(opts.date, { background_agents: 1 });
        }
      }
      if (opts?.subagentCollector) {
        opts.subagentCollector.recordSpawn(
          c.is_background_agent ? "background" : "foreground",
          (typeof toolInput === "object" ? (toolInput.subagent_type as string) : "") ?? "",
          (typeof toolInput === "object" ? (toolInput.description as string) : "") ?? "",
        );
      }
    }

    if (c.skill_name) {
      category = "skill";
      this.skillInvocations.set(c.skill_name, (this.skillInvocations.get(c.skill_name) ?? 0) + 1);
    }

    if (c.is_mcp) {
      category = "mcp";
      this._mcpCalls++;
      registerMcpTool(this.mcpServers, toolName);
      if (opts?.dailyTracker && opts.date) {
        opts.dailyTracker.add(opts.date, { mcp_calls: 1 });
      }
    }

    if (c.is_plan_enter && opts?.planTracker) {
      if (opts.timestampDt) {
        opts.planTracker.onEnter(opts.timestampDt);
      } else {
        opts.planTracker.entries++;
      }
    }

    if (c.is_plan_exit && opts?.planTracker) {
      if (opts.timestampDt) {
        opts.planTracker.onExit(opts.timestampDt);
      } else {
        opts.planTracker.exits++;
      }
    }

    this.toolCategories.set(toolName, category);
    return category;
  }

  get subagentCalls(): number { return this._subagentCalls; }
  set subagentCalls(v: number) { this._subagentCalls = v; }
  get backgroundAgents(): number { return this._backgroundAgents; }
  get mcpCalls(): number { return this._mcpCalls; }
  set mcpCalls(v: number) { this._mcpCalls = v; }
}

// --- SubagentCollector ---

export class SubagentCollector {
  private _metas: Partial<SubagentMeta>[] = [];

  recordSpawn(mode = "foreground", subagentType = "", description = ""): void {
    this._metas.push({
      ordinal: this._metas.length,
      mode,
      subagent_type: subagentType,
      description,
    });
  }

  enrich(idx: number, totalTokens = 0, totalToolCalls = 0): void {
    if (idx < this._metas.length) {
      this._metas[idx].total_tokens = totalTokens;
      this._metas[idx].total_tool_calls = totalToolCalls;
    }
  }

  ensureCount(minCount: number): void {
    while (this._metas.length < minCount) {
      this._metas.push({
        ordinal: this._metas.length,
        mode: "foreground",
        subagent_type: "",
        description: "",
      });
    }
  }

  finalize(): SubagentMeta[] {
    return this._metas.map((m) => createSubagentMeta(m));
  }

  get rawMetas(): Partial<SubagentMeta>[] { return this._metas; }
  get length(): number { return this._metas.length; }
}
